// @rspack/core/dist/logging/runtime.js
"use strict";
/**
 * The following code is modified based on
 * https://github.com/webpack/webpack/blob/4b4ca3b/lib/logging/runtime.js
 *
 * MIT Licensed
 * Author Tobias Koppers @sokra
 * Copyright (c) JS Foundation and other contributors
 * https://github.com/webpack/webpack/blob/main/LICENSE
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.hooks = exports.configureDefaultLogger = exports.getLogger = void 0;
const SyncBailHook = require("tapable/lib/SyncBailHook");
const Logger_1 = require("./Logger");
const createConsoleLogger_1 = require("./createConsoleLogger");
const currentDefaultLoggerOptions = {
    level: "info",
    debug: false,
    console
};
let currentDefaultLogger = (0, createConsoleLogger_1.createConsoleLogger)(currentDefaultLoggerOptions);
const getLogger = (name) => {
    return new Logger_1.Logger((type, args) => {
        if (exports.hooks.log.call(name, type, args) === undefined) {
            currentDefaultLogger(name, type, args);
        }
    }, childName => exports.getLogger(`${name}/${childName}`));
};
exports.getLogger = getLogger;
const configureDefaultLogger = (options) => {
    Object.assign(currentDefaultLoggerOptions, options);
    currentDefaultLogger = (0, createConsoleLogger_1.createConsoleLogger)(currentDefaultLoggerOptions);
};
exports.configureDefaultLogger = configureDefaultLogger;
exports.hooks = {
    log: new SyncBailHook(["origin", "type", "args"])
};
