// @rspack/core/dist/builtin-plugin/BannerPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BannerPlugin = void 0;
const binding_1 = require("@rspack/binding");
const zod_1 = require("../../compiled/zod");
const base_1 = require("./base");
const rule = zod_1.z.string().or(zod_1.z.instanceof(RegExp));
const rules = rule.or(rule.array());
const bannerFunction = zod_1.z
    .function()
    .args(zod_1.z.object({
    hash: zod_1.z.string(),
    chunk: zod_1.z.custom(),
    filename: zod_1.z.string()
}))
    .returns(zod_1.z.string());
const bannerContent = zod_1.z.string().or(bannerFunction);
const bannerPluginOptions = zod_1.z.strictObject({
    banner: bannerContent,
    entryOnly: zod_1.z.boolean().optional(),
    exclude: rules.optional(),
    include: rules.optional(),
    raw: zod_1.z.boolean().optional(),
    footer: zod_1.z.boolean().optional(),
    stage: zod_1.z.number().optional(),
    test: rules.optional()
});
const bannerPluginArgument = bannerContent.or(bannerPluginOptions);
exports.BannerPlugin = (0, base_1.create)(binding_1.BuiltinPluginName.BannerPlugin, (args) => {
    if (typeof args === "string" || typeof args === "function") {
        return {
            banner: args
        };
    }
    return {
        banner: args.banner,
        entryOnly: args.entryOnly,
        footer: args.footer,
        raw: args.raw,
        test: args.test,
        stage: args.stage,
        include: args.include,
        exclude: args.exclude
    };
});
