// @rspack/core/dist/builtin-plugin/EntryPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getRawEntryOptions = exports.EntryPlugin = void 0;
const binding_1 = require("@rspack/binding");
const config_1 = require("../config");
const util_1 = require("../util");
const base_1 = require("./base");
exports.EntryPlugin = (0, base_1.create)(binding_1.BuiltinPluginName.EntryPlugin, (context, entry, options = "") => {
    const entryOptions = typeof options === "string" ? { name: options } : options;
    return {
        context,
        entry,
        options: getRawEntryOptions(entryOptions)
    };
}, "make");
function getRawEntryOptions(entry) {
    const runtime = entry.runtime;
    const chunkLoading = entry.chunkLoading;
    return {
        name: entry.name,
        publicPath: entry.publicPath,
        baseUri: entry.baseUri,
        runtime,
        chunkLoading: !(0, util_1.isNil)(chunkLoading)
            ? (0, config_1.getRawChunkLoading)(chunkLoading)
            : undefined,
        asyncChunks: entry.asyncChunks,
        filename: entry.filename,
        library: entry.library && (0, config_1.getRawLibrary)(entry.library),
        layer: entry.layer ?? undefined,
        dependOn: entry.dependOn
    };
}
exports.getRawEntryOptions = getRawEntryOptions;
