// @rspack/core/dist/builtin-plugin/LightningCssMiminizerRspackPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.LightningCssMinimizerRspackPlugin = void 0;
const binding_1 = require("@rspack/binding");
const lightningcss_1 = require("../builtin-loader/lightningcss");
const base_1 = require("./base");
exports.LightningCssMinimizerRspackPlugin = (0, base_1.create)(binding_1.BuiltinPluginName.LightningCssMinimizerRspackPlugin, (options) => {
    const { include, exclude, draft, nonStandard, pseudoClasses } = options?.minimizerOptions ?? {};
    const targets = options?.minimizerOptions?.targets ?? "fully supports es6"; // last not support es module chrome version
    return {
        test: options?.test,
        include: options?.include,
        exclude: options?.exclude,
        removeUnusedLocalIdents: options?.removeUnusedLocalIdents ?? true,
        minimizerOptions: {
            errorRecovery: options?.minimizerOptions?.errorRecovery ?? true,
            unusedSymbols: options?.minimizerOptions?.unusedSymbols ?? [],
            include: include ? (0, lightningcss_1.toFeatures)(include) : undefined,
            exclude: exclude ? (0, lightningcss_1.toFeatures)(exclude) : undefined,
            targets: typeof targets === "string" ? [targets] : targets,
            draft: draft ? { customMedia: draft.customMedia ?? false } : undefined,
            nonStandard: nonStandard
                ? {
                    deepSelectorCombinator: nonStandard.deepSelectorCombinator ?? false
                }
                : undefined,
            pseudoClasses
        }
    };
});
