// @rspack/core/dist/exports.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.EvalDevToolModulePlugin = exports.EvalSourceMapDevToolPlugin = exports.SourceMapDevToolPlugin = exports.CopyRspackPlugin = exports.LightningCssMinimizerRspackPlugin = exports.SwcJsMinimizerRspackPlugin = exports.HtmlRspackPlugin = exports.sharing = exports.container = exports.optimize = exports.webworker = exports.javascript = exports.wasm = exports.library = exports.electron = exports.node = exports.web = exports.NormalModuleReplacementPlugin = exports.LoaderTargetPlugin = exports.LoaderOptionsPlugin = exports.EnvironmentPlugin = exports.NoEmitOnErrorsPlugin = exports.HotModuleReplacementPlugin = exports.ExternalsPlugin = exports.DynamicEntryPlugin = exports.EntryPlugin = exports.ProgressPlugin = exports.DefinePlugin = exports.ProvidePlugin = exports.IgnorePlugin = exports.BannerPlugin = exports.EntryOptionPlugin = exports.util = exports.ValidationError = exports.config = exports.sources = exports.WebpackError = exports.Template = exports.ModuleFilenameHelpers = exports.Stats = exports.RuntimeGlobals = exports.NormalModule = exports.MultiStats = exports.WebpackOptionsApply = exports.RspackOptionsApply = exports.MultiCompiler = exports.Compiler = exports.Compilation = exports.version = exports.rspackVersion = void 0;
exports.experiments = exports.CssExtractRspackPlugin = void 0;
const package_json_1 = require("../package.json");
// this is a hack to be compatible with plugin which detect webpack's version
const rspackVersion = package_json_1.version;
exports.rspackVersion = rspackVersion;
const version = package_json_1.webpackVersion;
exports.version = version;
var Compilation_1 = require("./Compilation");
Object.defineProperty(exports, "Compilation", { enumerable: true, get: function () { return Compilation_1.Compilation; } });
var Compiler_1 = require("./Compiler");
Object.defineProperty(exports, "Compiler", { enumerable: true, get: function () { return Compiler_1.Compiler; } });
var MultiCompiler_1 = require("./MultiCompiler");
Object.defineProperty(exports, "MultiCompiler", { enumerable: true, get: function () { return MultiCompiler_1.MultiCompiler; } });
const rspackOptionsApply_1 = require("./rspackOptionsApply");
Object.defineProperty(exports, "RspackOptionsApply", { enumerable: true, get: function () { return rspackOptionsApply_1.RspackOptionsApply; } });
Object.defineProperty(exports, "WebpackOptionsApply", { enumerable: true, get: function () { return rspackOptionsApply_1.RspackOptionsApply; } });
var MultiStats_1 = require("./MultiStats");
Object.defineProperty(exports, "MultiStats", { enumerable: true, get: function () { return MultiStats_1.MultiStats; } });
var NormalModule_1 = require("./NormalModule");
Object.defineProperty(exports, "NormalModule", { enumerable: true, get: function () { return NormalModule_1.NormalModule; } });
var RuntimeGlobals_1 = require("./RuntimeGlobals");
Object.defineProperty(exports, "RuntimeGlobals", { enumerable: true, get: function () { return RuntimeGlobals_1.RuntimeGlobals; } });
var Stats_1 = require("./Stats");
Object.defineProperty(exports, "Stats", { enumerable: true, get: function () { return Stats_1.Stats; } });
// API extractor not working with some re-exports, see: https://github.com/microsoft/fluentui/issues/20694
const ModuleFilenameHelpers = __importStar(require("./lib/ModuleFilenameHelpers"));
exports.ModuleFilenameHelpers = ModuleFilenameHelpers;
// API extractor not working with some re-exports, see: https://github.com/microsoft/fluentui/issues/20694
var Template_1 = require("./Template");
Object.defineProperty(exports, "Template", { enumerable: true, get: function () { return Template_1.Template; } });
exports.WebpackError = Error;
const sources = require("../compiled/webpack-sources");
exports.sources = sources;
const config_1 = require("./config");
exports.config = {
    getNormalizedRspackOptions: config_1.getNormalizedRspackOptions,
    applyRspackOptionsDefaults: config_1.applyRspackOptionsDefaults,
    getNormalizedWebpackOptions: config_1.getNormalizedRspackOptions,
    applyWebpackOptionsDefaults: config_1.applyRspackOptionsDefaults
};
const validate_1 = require("./util/validate");
Object.defineProperty(exports, "ValidationError", { enumerable: true, get: function () { return validate_1.ValidationError; } });
const cleverMerge_1 = require("./util/cleverMerge");
const createHash_1 = require("./util/createHash");
exports.util = { createHash: createHash_1.createHash, cleverMerge: cleverMerge_1.cachedCleverMerge };
var EntryOptionPlugin_1 = require("./lib/EntryOptionPlugin");
Object.defineProperty(exports, "EntryOptionPlugin", { enumerable: true, get: function () { return __importDefault(EntryOptionPlugin_1).default; } });
var builtin_plugin_1 = require("./builtin-plugin");
Object.defineProperty(exports, "BannerPlugin", { enumerable: true, get: function () { return builtin_plugin_1.BannerPlugin; } });
var builtin_plugin_2 = require("./builtin-plugin");
Object.defineProperty(exports, "IgnorePlugin", { enumerable: true, get: function () { return builtin_plugin_2.IgnorePlugin; } });
var builtin_plugin_3 = require("./builtin-plugin");
Object.defineProperty(exports, "ProvidePlugin", { enumerable: true, get: function () { return builtin_plugin_3.ProvidePlugin; } });
var builtin_plugin_4 = require("./builtin-plugin");
Object.defineProperty(exports, "DefinePlugin", { enumerable: true, get: function () { return builtin_plugin_4.DefinePlugin; } });
var builtin_plugin_5 = require("./builtin-plugin");
Object.defineProperty(exports, "ProgressPlugin", { enumerable: true, get: function () { return builtin_plugin_5.ProgressPlugin; } });
var builtin_plugin_6 = require("./builtin-plugin");
Object.defineProperty(exports, "EntryPlugin", { enumerable: true, get: function () { return builtin_plugin_6.EntryPlugin; } });
var builtin_plugin_7 = require("./builtin-plugin");
Object.defineProperty(exports, "DynamicEntryPlugin", { enumerable: true, get: function () { return builtin_plugin_7.DynamicEntryPlugin; } });
var builtin_plugin_8 = require("./builtin-plugin");
Object.defineProperty(exports, "ExternalsPlugin", { enumerable: true, get: function () { return builtin_plugin_8.ExternalsPlugin; } });
var builtin_plugin_9 = require("./builtin-plugin");
Object.defineProperty(exports, "HotModuleReplacementPlugin", { enumerable: true, get: function () { return builtin_plugin_9.HotModuleReplacementPlugin; } });
var builtin_plugin_10 = require("./builtin-plugin");
Object.defineProperty(exports, "NoEmitOnErrorsPlugin", { enumerable: true, get: function () { return builtin_plugin_10.NoEmitOnErrorsPlugin; } });
var EnvironmentPlugin_1 = require("./lib/EnvironmentPlugin");
Object.defineProperty(exports, "EnvironmentPlugin", { enumerable: true, get: function () { return EnvironmentPlugin_1.EnvironmentPlugin; } });
var LoaderOptionsPlugin_1 = require("./lib/LoaderOptionsPlugin");
Object.defineProperty(exports, "LoaderOptionsPlugin", { enumerable: true, get: function () { return LoaderOptionsPlugin_1.LoaderOptionsPlugin; } });
var LoaderTargetPlugin_1 = require("./lib/LoaderTargetPlugin");
Object.defineProperty(exports, "LoaderTargetPlugin", { enumerable: true, get: function () { return LoaderTargetPlugin_1.LoaderTargetPlugin; } });
var NormalModuleReplacementPlugin_1 = require("./lib/NormalModuleReplacementPlugin");
Object.defineProperty(exports, "NormalModuleReplacementPlugin", { enumerable: true, get: function () { return NormalModuleReplacementPlugin_1.NormalModuleReplacementPlugin; } });
const builtin_plugin_11 = require("./builtin-plugin");
exports.web = {
    FetchCompileAsyncWasmPlugin: builtin_plugin_11.FetchCompileAsyncWasmPlugin
};
const builtin_plugin_12 = require("./builtin-plugin");
const NodeEnvironmentPlugin_1 = __importDefault(require("./node/NodeEnvironmentPlugin"));
const NodeTemplatePlugin_1 = __importDefault(require("./node/NodeTemplatePlugin"));
exports.node = {
    NodeTargetPlugin: builtin_plugin_12.NodeTargetPlugin,
    NodeTemplatePlugin: NodeTemplatePlugin_1.default,
    NodeEnvironmentPlugin: NodeEnvironmentPlugin_1.default
};
const builtin_plugin_13 = require("./builtin-plugin");
exports.electron = { ElectronTargetPlugin: builtin_plugin_13.ElectronTargetPlugin };
const builtin_plugin_14 = require("./builtin-plugin");
exports.library = { EnableLibraryPlugin: builtin_plugin_14.EnableLibraryPlugin };
const builtin_plugin_15 = require("./builtin-plugin");
exports.wasm = { EnableWasmLoadingPlugin: builtin_plugin_15.EnableWasmLoadingPlugin };
const builtin_plugin_16 = require("./builtin-plugin");
exports.javascript = {
    EnableChunkLoadingPlugin: builtin_plugin_16.EnableChunkLoadingPlugin,
    JavascriptModulesPlugin: builtin_plugin_16.JavascriptModulesPlugin
};
const builtin_plugin_17 = require("./builtin-plugin");
exports.webworker = { WebWorkerTemplatePlugin: builtin_plugin_17.WebWorkerTemplatePlugin };
const builtin_plugin_18 = require("./builtin-plugin");
const builtin_plugin_19 = require("./builtin-plugin");
const builtin_plugin_20 = require("./builtin-plugin");
exports.optimize = {
    LimitChunkCountPlugin: builtin_plugin_18.LimitChunkCountPlugin,
    RuntimeChunkPlugin: builtin_plugin_19.RuntimeChunkPlugin,
    SplitChunksPlugin: builtin_plugin_20.SplitChunksPlugin
};
const ModuleFederationPlugin_1 = require("./container/ModuleFederationPlugin");
const ModuleFederationPluginV1_1 = require("./container/ModuleFederationPluginV1");
const ContainerPlugin_1 = require("./container/ContainerPlugin");
const ContainerReferencePlugin_1 = require("./container/ContainerReferencePlugin");
exports.container = {
    ContainerPlugin: ContainerPlugin_1.ContainerPlugin,
    ContainerReferencePlugin: ContainerReferencePlugin_1.ContainerReferencePlugin,
    ModuleFederationPlugin: ModuleFederationPlugin_1.ModuleFederationPlugin,
    ModuleFederationPluginV1: ModuleFederationPluginV1_1.ModuleFederationPluginV1
};
const ConsumeSharedPlugin_1 = require("./sharing/ConsumeSharedPlugin");
const ProvideSharedPlugin_1 = require("./sharing/ProvideSharedPlugin");
const SharePlugin_1 = require("./sharing/SharePlugin");
exports.sharing = {
    ProvideSharedPlugin: ProvideSharedPlugin_1.ProvideSharedPlugin,
    ConsumeSharedPlugin: ConsumeSharedPlugin_1.ConsumeSharedPlugin,
    SharePlugin: SharePlugin_1.SharePlugin
};
var builtin_plugin_21 = require("./builtin-plugin");
Object.defineProperty(exports, "HtmlRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_21.HtmlRspackPlugin; } });
var builtin_plugin_22 = require("./builtin-plugin");
Object.defineProperty(exports, "SwcJsMinimizerRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_22.SwcJsMinimizerRspackPlugin; } });
var builtin_plugin_23 = require("./builtin-plugin");
Object.defineProperty(exports, "LightningCssMinimizerRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_23.LightningCssMinimizerRspackPlugin; } });
var builtin_plugin_24 = require("./builtin-plugin");
Object.defineProperty(exports, "CopyRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_24.CopyRspackPlugin; } });
var builtin_plugin_25 = require("./builtin-plugin");
Object.defineProperty(exports, "SourceMapDevToolPlugin", { enumerable: true, get: function () { return builtin_plugin_25.SourceMapDevToolPlugin; } });
var builtin_plugin_26 = require("./builtin-plugin");
Object.defineProperty(exports, "EvalSourceMapDevToolPlugin", { enumerable: true, get: function () { return builtin_plugin_26.EvalSourceMapDevToolPlugin; } });
var builtin_plugin_27 = require("./builtin-plugin");
Object.defineProperty(exports, "EvalDevToolModulePlugin", { enumerable: true, get: function () { return builtin_plugin_27.EvalDevToolModulePlugin; } });
var builtin_plugin_28 = require("./builtin-plugin");
Object.defineProperty(exports, "CssExtractRspackPlugin", { enumerable: true, get: function () { return builtin_plugin_28.CssExtractRspackPlugin; } });
///// Experiments Stuff /////
const binding_1 = require("@rspack/binding");
exports.experiments = {
    globalTrace: {
        register: binding_1.registerGlobalTrace,
        cleanup: binding_1.cleanupGlobalTrace
    }
};
