// @rspack/core/dist/builtin-loader/lightningcss/index.js
"use strict";
// code modified based on https://github.com/parcel-bundler/lightningcss/blob/34b67a431c043fda5d4979bcdccb3008d082e243/node/browserslistToTargets.js
Object.defineProperty(exports, "__esModule", { value: true });
exports.Features = exports.toFeatures = exports.browserslistToTargets = void 0;
/**
MIT License

Copyright (c) 2021-present Devon Govett

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.
 */
const BROWSER_MAPPING = {
    and_chr: "chrome",
    and_ff: "firefox",
    ie_mob: "ie",
    op_mob: "opera",
    and_qq: null,
    and_uc: null,
    baidu: null,
    bb: null,
    kaios: null,
    op_mini: null
};
function browserslistToTargets(browserslist) {
    const targets = {};
    for (const browser of browserslist) {
        const [name, v] = browser.split(" ");
        if (BROWSER_MAPPING[name] === null) {
            continue;
        }
        const version = parseVersion(v);
        if (version == null) {
            continue;
        }
        if (targets[name] == null || version < targets[name]) {
            targets[name] = version;
        }
    }
    return targets;
}
exports.browserslistToTargets = browserslistToTargets;
function toFeatures(featureOptions) {
    let feature = 0;
    for (const key of Reflect.ownKeys(featureOptions)) {
        if (featureOptions[key] !== true) {
            continue;
        }
        switch (key) {
            case "nesting":
                feature |= Features.Nesting;
                break;
            case "notSelectorList":
                feature |= Features.NotSelectorList;
                break;
            case "dirSelector":
                feature |= Features.DirSelector;
                break;
            case "langSelectorList":
                feature |= Features.LangSelectorList;
                break;
            case "isSelector":
                feature |= Features.IsSelector;
                break;
            case "textDecorationThicknessPercent":
                feature |= Features.TextDecorationThicknessPercent;
                break;
            case "mediaIntervalSyntax":
                feature |= Features.MediaIntervalSyntax;
                break;
            case "mediaRangeSyntax":
                feature |= Features.MediaRangeSyntax;
                break;
            case "customMediaQueries":
                feature |= Features.CustomMediaQueries;
                break;
            case "clampFunction":
                feature |= Features.ClampFunction;
                break;
            case "colorFunction":
                feature |= Features.ColorFunction;
                break;
            case "oklabColors":
                feature |= Features.OklabColors;
                break;
            case "labColors":
                feature |= Features.LabColors;
                break;
            case "p3Colors":
                feature |= Features.P3Colors;
                break;
            case "hexAlphaColors":
                feature |= Features.HexAlphaColors;
                break;
            case "spaceSeparatedColorNotation":
                feature |= Features.SpaceSeparatedColorNotation;
                break;
            case "fontFamilySystemUi":
                feature |= Features.FontFamilySystemUi;
                break;
            case "doublePositionGradients":
                feature |= Features.DoublePositionGradients;
                break;
            case "vendorPrefixes":
                feature |= Features.VendorPrefixes;
                break;
            case "logicalProperties":
                feature |= Features.LogicalProperties;
                break;
            case "selectors":
                feature |= Features.Selectors;
                break;
            case "mediaQueries":
                feature |= Features.MediaQueries;
                break;
            case "color":
                feature |= Features.Color;
                break;
        }
    }
    return feature;
}
exports.toFeatures = toFeatures;
function parseVersion(version) {
    const [major, minor = 0, patch = 0] = version
        .split("-")[0]
        .split(".")
        .map(v => Number.parseInt(v, 10));
    if (Number.isNaN(major) || Number.isNaN(minor) || Number.isNaN(patch)) {
        return null;
    }
    return (major << 16) | (minor << 8) | patch;
}
var Features;
(function (Features) {
    Features[Features["Empty"] = 0] = "Empty";
    Features[Features["Nesting"] = 1] = "Nesting";
    Features[Features["NotSelectorList"] = 2] = "NotSelectorList";
    Features[Features["DirSelector"] = 4] = "DirSelector";
    Features[Features["LangSelectorList"] = 8] = "LangSelectorList";
    Features[Features["IsSelector"] = 16] = "IsSelector";
    Features[Features["TextDecorationThicknessPercent"] = 32] = "TextDecorationThicknessPercent";
    Features[Features["MediaIntervalSyntax"] = 64] = "MediaIntervalSyntax";
    Features[Features["MediaRangeSyntax"] = 128] = "MediaRangeSyntax";
    Features[Features["CustomMediaQueries"] = 256] = "CustomMediaQueries";
    Features[Features["ClampFunction"] = 512] = "ClampFunction";
    Features[Features["ColorFunction"] = 1024] = "ColorFunction";
    Features[Features["OklabColors"] = 2048] = "OklabColors";
    Features[Features["LabColors"] = 4096] = "LabColors";
    Features[Features["P3Colors"] = 8192] = "P3Colors";
    Features[Features["HexAlphaColors"] = 16384] = "HexAlphaColors";
    Features[Features["SpaceSeparatedColorNotation"] = 32768] = "SpaceSeparatedColorNotation";
    Features[Features["FontFamilySystemUi"] = 65536] = "FontFamilySystemUi";
    Features[Features["DoublePositionGradients"] = 131072] = "DoublePositionGradients";
    Features[Features["VendorPrefixes"] = 262144] = "VendorPrefixes";
    Features[Features["LogicalProperties"] = 524288] = "LogicalProperties";
    Features[Features["Selectors"] = 31] = "Selectors";
    Features[Features["MediaQueries"] = 448] = "MediaQueries";
    Features[Features["Color"] = 64512] = "Color";
})(Features = exports.Features || (exports.Features = {}));
